activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb

```diff
@@ -61,13 +61,13 @@ def perform_query(raw_connection, sql, binds, type_casted_binds, prepare:, notif
               end
               result
             else
-              if prepare
-                stmt = @statements[sql] ||= raw_connection.prepare(sql)
-              else
-                stmt = raw_connection.prepare(sql)
-              end
-
               begin
+                if prepare
+                  stmt = @statements[sql] ||= raw_connection.prepare(sql)
+                else
+                  stmt = raw_connection.prepare(sql)
+                end
```

Comment:
I don't get this part of the change. If `prepare` raised, then there's no `stmt` to close, so why move it inside the `begin`?

Reply:
I refactored the adapter a bit, but I think we should just drop that commit.